### ISSUE TITLE
can't mix with ipfs-cluster

### ISSUE THREAD

**Issue by paleozogt** [ipfs/go-ipfs#8065]
#### Version information:
v0.8.0

#### Description:

I'm writing some ipfs code that brings in go-ipfs and the ipfs-cluster api, but it fails to compile.

```
go build ./...
# github.com/ipfs/go-filestore
../../../go/pkg/mod/github.com/ipfs/go-filestore@v0.0.3/fsrefstore.go:81:14: undefined: dshelp.DsKeyToCid
../../../go/pkg/mod/github.com/ipfs/go-filestore@v0.0.3/fsrefstore.go:101:21: undefined: dshelp.CidToDsKey
../../../go/pkg/mod/github.com/ipfs/go-filestore@v0.0.3/fsrefstore.go:145:21: undefined: dshelp.CidToDsKey
../../../go/pkg/mod/github.com/ipfs/go-filestore@v0.0.3/fsrefstore.go:258:9: undefined: dshelp.CidToDsKey
../../../go/pkg/mod/github.com/ipfs/go-filestore@v0.0.3/fsrefstore.go:303:16: undefined: dshelp.CidToDsKey
../../../go/pkg/mod/github.com/ipfs/go-filestore@v0.0.3/util.go:158:12: undefined: dshelp.DsKeyToCid
../../../go/pkg/mod/github.com/ipfs/go-filestore@v0.0.3/util.go:211:18: undefined: dshelp.DsKeyToCid
```

Here's the `go.mod`:
```
module paleozogt/ipfs-test

go 1.12

require (
	github.com/ipfs/go-ipfs v0.8.0
	github.com/ipfs/ipfs-cluster v0.13.2
)
```

Here's the `main.go`:

```
package main

import (
	_ "github.com/ipfs/go-ipfs/plugin"
	_ "github.com/ipfs/ipfs-cluster"
)

func main() {
}
```

**Comment by gammazero** [comment]
The problem is caused by these modules requiring different versions of `go-ipfs-ds-help` that are not compatible:
- `ipfs-cluster v0.13.2` => `go-ipfs-ds-help v1.0.0` 
- `go-ipfs v0.8.0` => `go-filestore v0.0.3` => `go-ipfs-ds-help v0.1.1`

A solution, until `go-ipfs` updates its `go-filestore dependency`, is the execute the following commands in your repo:
```
go mod edit -replace=github.com/ipfs/go-filestore=github.com/ipfs/go-filestore@v1.0.0
go mod tidy
```
After that, your build should succeed.

There do not appear to be any breaking changes between go-filestore v0.0.3 and v1.0.0, so this should work.

**Comment by aschmahmann** [comment]
> There do not appear to be any breaking changes between go-filestore v0.0.3 and v1.0.0, so this should work.

There are no API breaking changes, but the way data is stored/retrieved from the filestore changed from CIDs to multihashes which is a breakage.

@hsanjuan any ideas/suggestions?

**Comment by mvdan** [comment]
> the way data is stored/retrieved from the filestore changed from CIDs to multihashes which is a breakage.

That's https://github.com/ipfs/go-ipfs/issues/6815, I think. Beware that the build will work, but CI will fail with some confusing sharness test errors - I was there three weeks ago, and it wasn't clear why stuff was breaking.

**Comment by aschmahmann** [comment]
Yes, that's #6815. Updating the filestore will pull in an updated version of the blockstore and you will break compatibility with any existing go-ipfs datastores.

Until #6815 is resolved I think any compatibility would need to be resolved by downgrading the version of go-ipfs-ds-help in your combined application since from what I can tell none of the features of go-ipfs-ds-help v1 are actually needed by ipfs-cluster. Using a replace directive in your go.mod for `go-ipfs-ds-help v0.1.1` will probably do the job.

However, ipfs-cluster might want to consider downgrading their version of go-ipfs-ds-help until we can resolve #6815, but @hsanjuan can give more context.

**Comment by hsanjuan** [comment]
@paleozogt

* As mentioned, upgrading the filestore+blockstore version used by go-ipfs will result in keys being stored differently in the datastore (addressed by multihash instead of cid). A vanilla v0.8.0 instance will not know how to read blocks from such local datastore. Depending on what you are doing, this may be acceptable. IPFS would otherwise work, it is just a matter of how data is stored locally and whether you want it to be compatible with current upstream.

* Cluster could potentially downgrade to make life easier, but this has implications down the dependency chain (ipfs-lite uses the v1 blockstore). Honestly, I don't want to go down this rabbit hole and I think it would be a breaking change. You could however fork your way down and downgrade.

I was thinking that patching go-ipfs-ds-help to add the missing methods back would fix things, but I think go dependency resolution will still end up using blockstore v1, so your blocks will be still addressed in the new way.

tl;dr: the go-mod-replace directive provided above is your best shot if you need to require both go-ipfs and cluster.

**Comment by BigLep** [comment]
We'll resolve this once the underlying linked issues get resolved.
